File: clients/lightroom-classic/bridge/bridge-paths.js
const os = require('os')
const path = require('path')

function createBridgePaths(options = {}) {
  const env = options.env || process.env
  const fsRoot = resolveRoot(env.TONEPILOT_LIGHTROOM_BRIDGE_ROOT || path.join(os.homedir(), '.tonepilot-lightroom-bridge'))
  const lightroomRoot = trimTrailingSeparator(env.TONEPILOT_LIGHTROOM_BRIDGE_LIGHTROOM_ROOT || fsRoot)

  return {
    fsRoot,
    lightroomRoot,
    fs: (...segments) => path.join(fsRoot, ...segments),
    lightroom: (...segments) => joinLightroomPath(lightroomRoot, segments)
  }
}

function resolveRoot(value) {
  const text = String(value || '')
  if (usesWindowsSeparator(text)) {
    return trimTrailingSeparator(text)
  }
  return path.resolve(text)
}

function joinLightroomPath(root, segments) {
  const normalized = segments
    .filter((segment) => segment !== undefined && segment !== null && String(segment).length > 0)
    .map((segment) => String(segment).replace(/^[/\\]+|[/\\]+$/g, ''))

  if (usesWindowsSeparator(root)) {
    return [trimTrailingSeparator(root), ...normalized].join('\\')
  }
  return path.join(root, ...normalized)
}

function trimTrailingSeparator(value) {
  const text = String(value || '')
  if (/^[a-zA-Z]:[\\/]?$/.test(text)) {
    return text.replace('/', '\\')
  }
  return text.replace(/[/\\]+$/g, '')
}

function usesWindowsSeparator(value) {
  return /^[a-zA-Z]:[\\/]/.test(value) || String(value).includes('\\')
}

module.exports = {
  createBridgePaths
}

File: clients/lightroom-classic/local-runtime/src/main/java/com/tonepilot/runtime/bridge/BridgePaths.java
package com.tonepilot.runtime.bridge;

import com.tonepilot.runtime.config.RuntimeProperties;

import java.nio.file.Path;

public class BridgePaths {

    private final Path fsRoot;
    private final String lightroomRoot;

    public BridgePaths(RuntimeProperties properties) {
        this(Path.of(properties.getBridge().getRoot()), properties.getBridge().getLightroomRoot());
    }

    public BridgePaths(Path fsRoot, String lightroomRoot) {
        this.fsRoot = fsRoot;
        this.lightroomRoot = trimTrailingSeparator(lightroomRoot == null || lightroomRoot.isBlank()
                ? fsRoot.toString()
                : lightroomRoot);
    }

    public Path fs(String first, String... more) {
        return fsRoot.resolve(Path.of(first, more));
    }

    public String lightroom(String... segments) {
        String separator = usesWindowsSeparator(lightroomRoot) ? "\\" : "/";
        StringBuilder builder = new StringBuilder(lightroomRoot);
        for (String segment : segments) {
            if (segment == null || segment.isBlank()) {
                continue;
            }
            builder.append(separator).append(segment.replaceAll("^[\\\\/]+|[\\\\/]+$", ""));
        }
        return builder.toString();
    }

    public Path fsRoot() {
        return fsRoot;
    }

    public String lightroomRoot() {
        return lightroomRoot;
    }

    private boolean usesWindowsSeparator(String value) {
        return value.matches("^[a-zA-Z]:[\\\\/].*") || value.contains("\\");
    }

    private String trimTrailingSeparator(String value) {
        if (value.matches("^[a-zA-Z]:[\\\\/]?$")) {
            return value.replace("/", "\\");
        }
        return value.replaceAll("[\\\\/]+$", "");
    }
}
